source/common/api/posix/os_sys_calls_impl.cc

```diff
@@ -149,32 +150,46 @@ bool OsSysCallsImpl::supportsIpTransparent() const {
   // https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/net/ipv6/ipv6_sockglue.c?id=6c46862280c5f55eda7750391bc65cd7e08c7535
   //
   // So, almost recent linux kernel supports both IP_TRANSPARENT and IPV6_TRANSPARENT options.
+  // But there are also has ipv4 only or ipv6 only scenarios.
   //
   // And these socket options need CAP_NET_ADMIN capability to be applied.
   // The CAP_NET_ADMIN capability should be applied by root user before call this function.
-  static const bool is_supported = [] {
+
+  int fd = 0;
+  int val = 0;
+  bool result = false;
+  bool is_support = false;
+  switch (ipVersion) {
+  case Network::Address::IpVersion::v4:
     // Check ipv4 case
-    int fd = ::socket(AF_INET, SOCK_DGRAM | SOCK_NONBLOCK, IPPROTO_UDP);
-    if (fd < 0) {
-      return false;
-    }
-    int val = 1;
-    bool result = (0 == ::setsockopt(fd, IPPROTO_IP, IP_TRANSPARENT, &val, sizeof(val)));
-    ::close(fd);
-    if (!result) {
-      return false;
-    }
+    static const bool ipv4_is_supported = [&] {
+      fd = ::socket(AF_INET, SOCK_DGRAM | SOCK_NONBLOCK, IPPROTO_UDP);
+      if (fd < 0) {
+        return false;
+      }
+      val = 1;
+      result = (0 == ::setsockopt(fd, IPPROTO_IP, IP_TRANSPARENT, &val, sizeof(val)));
+      ::close(fd);
+      return result;
+    }();
```

Comment:
Could we reuse this lambda for ipv4 and ipv6? Seems they have almost same logic?

ideally, the logic could be something like following pseudo-code:
```
static constexpr transparent_supported = [] (..., ...) {};

static bool ipv4_is_supported = transparent_supported(AF_INET, IPPROTO_IP);
static bool ipv6_is_supported = transparent_supported(AF_INET6, IPPROTO_IPV6);

// switch is also ok for me.
return ip_version == Network::Address::IpVersion::v4 ? ipv4_is_supported : ipv6_is_supported;

```